src/material/form-field/form-field.ts

```diff
@@ -703,12 +703,18 @@ export class MatFormField
     const element: HTMLElement = this._elementRef.nativeElement;
     if (element.getRootNode) {
       const rootNode = element.getRootNode();
-      // If the element is inside the DOM the root node will be either the document
-      // or the closest shadow root, otherwise it'll be the element itself.
-      return rootNode && rootNode !== element;
+      // If the element is inside the DOM the root node will be either the document,
+      // the closest shadow root or an element that is not yet rendered, otherwise it'll be the element itself.
+      if (rootNode && rootNode !== element && rootNode === document) {
```

Comment:
I just tried it out and it breaks prefixes in the shadow DOM. I think we can simplify the entire method to this:

```
private _isAttachedToDom(): boolean {
  const element: HTMLElement = this._elementRef.nativeElement;
  const rootNode = element.getRootNode();
  // If the element is inside the DOM the root node will be either the document,
  // the closest shadow root or an element that is not yet rendered, otherwise it'll be the element itself.
  return (
    rootNode &&
    rootNode !== element &&
    (rootNode === document || rootNode === _getShadowRoot(element))
  );
}
```

The fallback logic for browsers that don't support `getRootNode` isn't necessary anymore.

Reply:
looks way better this way 👍